Include Mono version in corlib version

akoeplinger opened the pull request:
This makes it possible to version the runtime<->corlib interface for each version branch independently and fixes the problem of monolite getting overwritten by changes in different branches when corlib version is still the same.

*Note: I pushed this to 2017-02 already to fix the build there but happy to incorporate feedback from this PR. The PR build will fail due to missing monolite, that's expected.*

marek-safar commented on mono/metadata/appdomain.c:
```diff
@@ -86,8 +86,16 @@
+ *    to be bumped too. The corlib version can be reset to 00 in that case.
  */
-#define MONO_CORLIB_VERSION 164
+#define MONO_CORLIB_VERSION 104090100
```
I think we should keep at least 3 digits available for the api bump also not sure we need 10 prefix.
In summary thinking about something like `40901000

lewurm replied:
the prefix is `1`; `04` is the major version. I think the `1` prefix is there to avoid truncation.

marek-safar replied:
Sure but why do we need 0 prefix there?

lewurm replied:
eventually, we might reach version >10 and then we don't want to fix up all places that were assuming major is only one digit 🙂

marek-safar replied:
We still need to fix all the places because major can be only single digit

akoeplinger replied:
The `1` prefix is there because in C an integer literal that starts with 0 is defined as octal, which we don't want.

The reason why I think two digits for the corlib version are enough is that we were at `164` before and this mechanism was introduced many years ago. Given that the number will be reset to 00 when we bump the Mono version this should be plenty enough.

And yes, the reason for major being two digits too is that we could end up with two digit major (see the 17.02 discussion).

lewurm commented:
build

kumpera commented on configure.ac:
```diff
@@ -29,6 +29,18 @@ AC_PROG_LN_S
+MONO_CORLIB_VER_BUILD=`echo $MONO_CORLIB_VER | cut -c 6,7 | sed 's/^0//'`
+MONO_CORLIB_VER_STRING=$MONO_CORLIB_VER_MAJOR.$MONO_CORLIB_VER_MINOR.$MONO_CORLIB_VER_BUILD
+
+if test "$VERSION" != "$MONO_CORLIB_VER_STRING"; then
```
This is the wrong approach.

configure should compute the suffix and both appdomain.c and Environment.cs  should use it.

akoeplinger commented:
@kumpera @marek-safar I updated the PR and incorporated what we discussed on Slack yesterday.

corlib version generation now happens automatically in configure.ac so you don't need to change multiple files when bumping Mono's version.

kumpera commented:
LGTM

vargaz commented on mcs/class/corlib/System/Environment.cs:
```diff
@@ -48,16 +48,11 @@ namespace System {
-		 * pair of this variable in the runtime in metadata/appdomain.c.
-		 * Changes which are already detected at runtime, like the addition
-		 * of icalls, do not require an increment.
+		 * This is the version number of the corlib-runtime interface.
```
The 'Changes which are already detected at runtime, like the addition
		 * of icalls, do not require an increment.' part is still true.

akoeplinger replied:
kumpera asked me to remove it

kumpera replied:
Hi Zoltan,

We decided to lose the shackles of decisions made in the 2000-2004 timeframe and give the corlib version a meaning more useful for our current reality.

Starting from this PR, the corlib version should be fully usable by monolite, so it must reflect changes that would break monolite as well.
